File: core/src/main/java/org/kiwix/kiwixmobile/core/settings/CorePrefsFragment.java
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.kiwix.kiwixmobile.core.settings;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.Settings;
import android.view.LayoutInflater;
import android.webkit.WebView;
import androidx.annotation.RequiresApi;
import androidx.core.content.ContextCompat;
import androidx.navigation.NavController;
import androidx.preference.EditTextPreference;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceFragmentCompat;
import com.google.android.material.snackbar.Snackbar;
import eu.mhutti1.utils.storage.StorageDevice;
import eu.mhutti1.utils.storage.StorageSelectDialog;
import java.io.File;
import java.util.List;
import java.util.Locale;
import javax.inject.Inject;
import kotlin.Unit;
import kotlin.io.FilesKt;
import org.jetbrains.annotations.NotNull;
import org.kiwix.kiwixmobile.core.CoreApp;
import org.kiwix.kiwixmobile.core.NightModeConfig;
import org.kiwix.kiwixmobile.core.R;
import org.kiwix.kiwixmobile.core.main.AddNoteDialog;
import org.kiwix.kiwixmobile.core.main.CoreMainActivity;
import org.kiwix.kiwixmobile.core.utils.LanguageUtils;
import org.kiwix.kiwixmobile.core.utils.SharedPreferenceUtil;
import org.kiwix.kiwixmobile.core.utils.dialog.DialogShower;
import org.kiwix.kiwixmobile.core.utils.dialog.KiwixDialog;

import static org.kiwix.kiwixmobile.core.utils.SharedPreferenceUtil.PREF_NIGHT_MODE;
import static org.kiwix.kiwixmobile.core.utils.SharedPreferenceUtil.PREF_STORAGE;

public abstract class CorePrefsFragment extends PreferenceFragmentCompat implements
  SettingsContract.View,
  SharedPreferences.OnSharedPreferenceChangeListener {

  public static final String PREF_VERSION = "pref_version";
  public static final String PREF_CLEAR_ALL_HISTORY = "pref_clear_all_history";
  public static final String PREF_CLEAR_ALL_NOTES = "pref_clear_all_notes";
  public static final String PREF_CREDITS = "pref_credits";
  public static final String PREF_MANAGE_EXTERNAL_STORAGE_PERMISSION =
    "pref_manage_external_storage";
  public static final String PREF_PERMISSION = "pref_permissions";
  private static final int ZOOM_OFFSET = 2;
  private static final int ZOOM_SCALE = 25;
  private static final String INTERNAL_TEXT_ZOOM = "text_zoom";
  @Inject
  SettingsPresenter presenter;
  @Inject
  protected SharedPreferenceUtil sharedPreferenceUtil;
  @Inject
  protected StorageCalculator storageCalculator;
  @Inject
  protected NightModeConfig nightModeConfig;
  @Inject
  protected DialogShower alertDialogShower;

  @Override
  public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
    CoreApp.getCoreComponent()
      .activityComponentBuilder()
      .activity(getActivity())
      .build()
      .inject(this);
    addPreferencesFromResource(R.xml.preferences);
    setStorage();
    setUpSettings();
    setupZoom();
    setMangeExternalStoragePermission();
    new LanguageUtils(getActivity()).changeFont(getActivity().getLayoutInflater(),
      sharedPreferenceUtil);
  }

  private void setupZoom() {
    final Preference textZoom = findPreference(INTERNAL_TEXT_ZOOM);
    textZoom.setOnPreferenceChangeListener(
      (preference, newValue) -> {
        sharedPreferenceUtil.setTextZoom((((Integer) newValue) + ZOOM_OFFSET) * ZOOM_SCALE);
        updateTextZoomSummary(textZoom);
        return true;
      });
    updateTextZoomSummary(textZoom);
  }

  private void updateTextZoomSummary(Preference textZoom) {
    textZoom.setSummary(getString(R.string.percentage, sharedPreferenceUtil.getTextZoom()));
  }

  protected abstract void setStorage();

  @Override
  public void onResume() {
    super.onResume();
    getPreferenceScreen().getSharedPreferences()
      .registerOnSharedPreferenceChangeListener(this);
  }

  @Override
  public void onPause() {
    super.onPause();
    getPreferenceScreen().getSharedPreferences()
      .unregisterOnSharedPreferenceChangeListener(this);
  }

  public void setUpSettings() {
    setAppVersionNumber();
  }

  protected void setUpLanguageChooser(String preferenceId) {
    ListPreference languagePref = findPreference(preferenceId);
    List<String> languageCodeList = new LanguageUtils(getActivity()).getKeys();
    languageCodeList.add(0, Locale.ROOT.getLanguage());
    final String selectedLang =
      selectedLanguage(languageCodeList, sharedPreferenceUtil.getPrefLanguage());
    languagePref.setEntries(languageDisplayValues(languageCodeList));
    languagePref.setEntryValues(languageCodeList.toArray(new String[0]));
    languagePref.setDefaultValue(selectedLang);
    languagePref.setValue(selectedLang);
    languagePref.setTitle(selectedLang.equals(Locale.ROOT.toString())
      ? getString(R.string.device_default)
      : new Locale(selectedLang).getDisplayLanguage());
    languagePref.setOnPreferenceChangeListener((preference, newValue) -> {
      String languageCode = (String) newValue;
      LanguageUtils.handleLocaleChange(getActivity(), languageCode);
      sharedPreferenceUtil.putPrefLanguage(languageCode);
      restartActivity();
      return true;
    });
  }

  private void restartActivity() {
    final CoreMainActivity activity = (CoreMainActivity) getActivity();
    final NavController navController = activity.getNavController();
    navController.popBackStack();
    navController.navigate(activity.getSettingsFragmentResId());
  }

  @NotNull private String selectedLanguage(List<String> languageCodeList, String langPref) {
    return languageCodeList.contains(langPref) ? langPref : "en";
  }

  @NotNull private String[] languageDisplayValues(List<String> languageCodeList) {
    String[] entries = new String[languageCodeList.size()];
    entries[0] = getString(R.string.device_default);
    for (int i = 1; i < languageCodeList.size(); i++) {
      Locale locale = new Locale(languageCodeList.get(i));
      entries[i] = locale.getDisplayLanguage() + " (" + locale.getDisplayLanguage(locale) + ") ";
    }
    return entries;
  }

  private void setAppVersionNumber() {
    EditTextPreference versionPref = findPreference(PREF_VERSION);
    versionPref.setSummary(getVersionName() + " Build: " + getVersionCode());
  }

  private void setMangeExternalStoragePermission() {
    Preference permissionPref = findPreference(PREF_MANAGE_EXTERNAL_STORAGE_PERMISSION);
    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
      showPermissionPreference();
      boolean externalStorageManager = Environment.isExternalStorageManager();
      if (externalStorageManager) {
        permissionPref.setSummary(R.string.allowed);
      } else {
        permissionPref.setSummary(R.string.not_allowed);
      }
      permissionPref.setOnPreferenceClickListener(
        preference -> {
            navigateToSettings();
            return true;
        });
    }
  }
  private void showPermissionPreference() {
    PreferenceCategory preferenceCategory = findPreference(PREF_PERMISSION);
    preferenceCategory.setVisible(true);
  }

  private int getVersionCode() {
    try {
      return getActivity().getPackageManager()
        .getPackageInfo(getActivity().getPackageName(), 0).versionCode;
    } catch (PackageManager.NameNotFoundException e) {
      throw new RuntimeException(e);
    }
  }

  private String getVersionName() {
    try {
      return getActivity().getPackageManager()
        .getPackageInfo(getActivity().getPackageName(), 0).versionName;
    } catch (PackageManager.NameNotFoundException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
    if (key.equals(PREF_NIGHT_MODE)) {
      sharedPreferenceUtil.updateNightMode();
    }
  }

  private void clearAllHistoryDialog() {
    alertDialogShower.show(KiwixDialog.ClearAllHistory.INSTANCE, () -> {
      presenter.clearHistory();
      Snackbar.make(getView(), R.string.all_history_cleared, Snackbar.LENGTH_SHORT).show();
      return Unit.INSTANCE;
    });
  }

  private void showClearAllNotesDialog() {
    alertDialogShower.show(KiwixDialog.ClearAllNotes.INSTANCE, () -> {
      clearAllNotes();
      return Unit.INSTANCE;
    });
  }

  private void clearAllNotes() {
    if (CoreApp.getInstance().isExternalStorageWritable()) {
      if (ContextCompat.checkSelfPermission(getActivity(),
        Manifest.permission.WRITE_EXTERNAL_STORAGE)
        != PackageManager.PERMISSION_GRANTED) {
        Snackbar.make(getView(), R.string.ext_storage_permission_not_granted, Snackbar.LENGTH_SHORT)
          .show();
        return;
      }

      if (FilesKt.deleteRecursively(new File(AddNoteDialog.NOTES_DIRECTORY))) {
        Snackbar.make(getView(), R.string.notes_deletion_successful, Snackbar.LENGTH_SHORT).show();
        return;
      }
    }
    Snackbar.make(getView(), R.string.notes_deletion_unsuccessful, Snackbar.LENGTH_SHORT).show();
  }

  @SuppressLint("SetJavaScriptEnabled")
  public void openCredits() {
    @SuppressLint("InflateParams") WebView view =
      (WebView) LayoutInflater.from(getActivity()).inflate(R.layout.credits_webview, null);
    view.loadUrl("file:///android_asset/credits.html");
    if (nightModeConfig.isNightModeActive()) {
      view.getSettings().setJavaScriptEnabled(true);
      view.setBackgroundColor(0);
    }
    alertDialogShower.show(new KiwixDialog.OpenCredits(() -> view));
  }

  @Override public boolean onPreferenceTreeClick(Preference preference) {
    if (preference.getKey().equalsIgnoreCase(PREF_CLEAR_ALL_HISTORY)) {
      clearAllHistoryDialog();
    }
    if (preference.getKey().equalsIgnoreCase(PREF_CLEAR_ALL_NOTES)) {
      showClearAllNotesDialog();
    }

    if (preference.getKey().equalsIgnoreCase(PREF_CREDITS)) {
      openCredits();
    }

    if (preference.getKey().equalsIgnoreCase(PREF_STORAGE)) {
      openFolderSelect();
    }

    return true;
  }

  public void openFolderSelect() {
    StorageSelectDialog dialogFragment = new StorageSelectDialog();
    dialogFragment.setOnSelectAction(this::onStorageDeviceSelected);
    dialogFragment.show(getActivity().getSupportFragmentManager(),
      getResources().getString(R.string.pref_storage));
  }

  private Unit onStorageDeviceSelected(StorageDevice storageDevice) {
    findPreference(PREF_STORAGE).setSummary(
      storageCalculator.calculateAvailableSpace(storageDevice.getFile())
    );
    sharedPreferenceUtil.putPrefStorage(storageDevice.getName());
    if (storageDevice.isInternal()) {
      findPreference(PREF_STORAGE).setTitle(getString(R.string.internal_storage));
    } else {
      findPreference(PREF_STORAGE).setTitle(getString(R.string.external_storage));
    }
    return Unit.INSTANCE;
  }

  // TODO: 28/10/21 Refactor the code. We are using it twice.
  @RequiresApi(Build.VERSION_CODES.R)
  private void navigateToSettings() {
    Intent intent = new Intent();
    intent.setAction(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION);
    Uri uri = Uri.fromParts("package", getActivity().getPackageName(), null);
    intent.setData(uri);
    startActivity(intent);
  }
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/settings/CorePrefsFragment.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.settings

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity.RESULT_OK
import android.content.ActivityNotFoundException
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.webkit.WebView
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import androidx.preference.EditTextPreference
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.google.android.material.snackbar.Snackbar
import eu.mhutti1.utils.storage.StorageDevice
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.kiwix.kiwixmobile.core.CoreApp.Companion.coreComponent
import org.kiwix.kiwixmobile.core.CoreApp.Companion.instance
import org.kiwix.kiwixmobile.core.DarkModeConfig
import org.kiwix.kiwixmobile.core.R
import org.kiwix.kiwixmobile.core.compat.CompatHelper.Companion.getPackageInformation
import org.kiwix.kiwixmobile.core.compat.CompatHelper.Companion.getVersionCode
import org.kiwix.kiwixmobile.core.dao.LibkiwixBookmarks
import org.kiwix.kiwixmobile.core.extensions.toast
import org.kiwix.kiwixmobile.core.main.AddNoteDialog
import org.kiwix.kiwixmobile.core.main.CoreMainActivity
import org.kiwix.kiwixmobile.core.navigateToAppSettings
import org.kiwix.kiwixmobile.core.utils.EXTERNAL_SELECT_POSITION
import org.kiwix.kiwixmobile.core.utils.INTERNAL_SELECT_POSITION
import org.kiwix.kiwixmobile.core.utils.LanguageUtils
import org.kiwix.kiwixmobile.core.utils.LanguageUtils.Companion.handleLocaleChange
import org.kiwix.kiwixmobile.core.utils.SharedPreferenceUtil
import org.kiwix.kiwixmobile.core.utils.dialog.DialogShower
import org.kiwix.kiwixmobile.core.utils.dialog.KiwixDialog
import org.kiwix.kiwixmobile.core.utils.dialog.KiwixDialog.OpenCredits
import java.io.File
import java.io.InputStream
import java.util.Locale
import javax.inject.Inject
import javax.xml.parsers.DocumentBuilderFactory

abstract class CorePrefsFragment :
  PreferenceFragmentCompat(),
  SettingsContract.View,
  OnSharedPreferenceChangeListener {

  @JvmField
  @Inject
  internal var presenter: SettingsPresenter? = null

  @JvmField
  @Inject
  protected var sharedPreferenceUtil: SharedPreferenceUtil? = null

  @JvmField
  @Inject
  protected var storageCalculator: StorageCalculator? = null

  @JvmField
  @Inject
  protected var darkModeConfig: DarkModeConfig? = null

  @JvmField
  @Inject
  protected var alertDialogShower: DialogShower? = null

  @JvmField
  @Inject
  internal var libkiwixBookmarks: LibkiwixBookmarks? = null
  override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
    lifecycleScope.launch {
      coreComponent
        .activityComponentBuilder()
        .activity(requireActivity())
        .build()
        .inject(this@CorePrefsFragment)
      addPreferencesFromResource(R.xml.preferences)
      setStorage()
      setUpSettings()
      setupZoom()
      sharedPreferenceUtil?.let {
        LanguageUtils(requireActivity()).changeFont(
          requireActivity(),
          it
        )
      }
    }
  }

  private fun setupZoom() {
    val textZoom = findPreference<Preference>(INTERNAL_TEXT_ZOOM)
    textZoom?.onPreferenceChangeListener =
      Preference.OnPreferenceChangeListener { _, newValue ->
        sharedPreferenceUtil?.textZoom = (newValue as Int + ZOOM_OFFSET) * ZOOM_SCALE
        updateTextZoomSummary(textZoom)
        true
      }
    updateTextZoomSummary(textZoom)
  }

  private fun updateTextZoomSummary(textZoom: Preference?) {
    textZoom?.summary = getString(R.string.percentage, sharedPreferenceUtil?.textZoom)
  }

  protected abstract suspend fun setStorage()
  override fun onResume() {
    super.onResume()
    preferenceScreen.sharedPreferences
      ?.registerOnSharedPreferenceChangeListener(this)
  }

  override fun onPause() {
    super.onPause()
    preferenceScreen.sharedPreferences
      ?.unregisterOnSharedPreferenceChangeListener(this)
  }

  override fun onDestroyView() {
    presenter?.dispose()
    storagePermissionForNotesLauncher?.unregister()
    storagePermissionForNotesLauncher = null
    super.onDestroyView()
  }

  private fun setUpSettings() {
    setAppVersionNumber()
  }

  protected fun setUpLanguageChooser(preferenceId: String) {
    findPreference<ListPreference>(
      preferenceId
    )?.let { languagePref ->
      var languageCodeList = LanguageUtils(requireActivity()).keys
      languageCodeList = listOf(Locale.ROOT.language) + languageCodeList
      sharedPreferenceUtil?.let { sharedPreferenceUtil ->
        val selectedLang = selectedLanguage(languageCodeList, sharedPreferenceUtil.prefLanguage)
        languagePref.entries = languageDisplayValues(languageCodeList)
        languagePref.entryValues = languageCodeList.toTypedArray()
        languagePref.setDefaultValue(selectedLang)
        languagePref.value = selectedLang
        languagePref.title =
          if (selectedLang == Locale.ROOT.toString())
            getString(R.string.device_default)
          else Locale(
            selectedLang
          ).displayLanguage
        languagePref.onPreferenceChangeListener =
          Preference.OnPreferenceChangeListener { _, newValue ->
            val languageCode = newValue as String?
            languageCode?.let {
              handleLocaleChange(requireActivity(), it, sharedPreferenceUtil)
              sharedPreferenceUtil.putPrefLanguage(it)
              restartActivity()
            }
            true
          }
      }
    }
  }

  private fun restartActivity() {
    (activity as CoreMainActivity?)?.let {
      it.navController.apply {
        popBackStack()
        navigate(it.settingsFragmentResId)
      }
    }
  }

  private fun selectedLanguage(
    languageCodeList: List<String>,
    langPref: String
  ): String =
    if (languageCodeList.contains(langPref)) langPref else "en"

  private fun languageDisplayValues(languageCodeList: List<String>): Array<String?> {
    val entries = arrayOfNulls<String>(languageCodeList.size)
    entries[0] = getString(R.string.device_default)
    for (i in 1 until languageCodeList.size) {
      val locale = Locale(languageCodeList[i])
      entries[i] = locale.displayLanguage + " (" + locale.getDisplayLanguage(locale) + ") "
    }
    return entries
  }

  private fun setAppVersionNumber() {
    findPreference<EditTextPreference>(PREF_VERSION)?.apply {
      summary = "$versionName Build: $versionCode"
    }
  }

  private val versionCode: Int
    @Suppress("TooGenericExceptionThrown")
    get() = try {
      requireActivity().packageManager
        .getPackageInformation(requireActivity().packageName, 0).getVersionCode()
    } catch (e: PackageManager.NameNotFoundException) {
      throw RuntimeException(e)
    }
  private val versionName: String
    @Suppress("TooGenericExceptionThrown")
    get() = try {
      requireActivity().packageManager
        .getPackageInformation(requireActivity().packageName, 0).versionName
    } catch (e: PackageManager.NameNotFoundException) {
      throw RuntimeException(e)
    }

  override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
    if (key == SharedPreferenceUtil.PREF_DARK_MODE) {
      sharedPreferenceUtil?.updateDarkMode()
    }
  }

  private fun clearAllHistoryDialog() {
    alertDialogShower?.show(KiwixDialog.ClearAllHistory, {
      presenter?.clearHistory()
      Snackbar.make(requireView(), R.string.all_history_cleared, Snackbar.LENGTH_SHORT).show()
    })
  }

  private fun showClearAllNotesDialog() {
    alertDialogShower?.show(KiwixDialog.ClearAllNotes, ::clearAllNotes)
  }

  private fun clearAllNotes() {
    if (instance.isExternalStorageWritable) {
      if (ContextCompat.checkSelfPermission(
          requireActivity(),
          Manifest.permission.WRITE_EXTERNAL_STORAGE
        )
        != PackageManager.PERMISSION_GRANTED &&
        sharedPreferenceUtil?.isPlayStoreBuildWithAndroid11OrAbove() == false &&
        Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU
      ) {
        Snackbar.make(
          requireView(),
          R.string.ext_storage_permission_not_granted,
          Snackbar.LENGTH_SHORT
        )
          .show()
        return
      }
      if (File(AddNoteDialog.NOTES_DIRECTORY).deleteRecursively()) {
        Snackbar.make(requireView(), R.string.notes_deletion_successful, Snackbar.LENGTH_SHORT)
          .show()
        return
      }
    }
    Snackbar.make(requireView(), R.string.notes_deletion_unsuccessful, Snackbar.LENGTH_SHORT).show()
  }

  @SuppressLint("SetJavaScriptEnabled")
  fun openCredits() {
    @SuppressLint("InflateParams") val view =
      LayoutInflater.from(
        requireActivity()
      ).inflate(R.layout.credits_webview, null) as WebView
    view.loadUrl("file:///android_asset/credits.html")
    if (darkModeConfig?.isDarkModeActive() == true) {
      view.settings.javaScriptEnabled = true
      view.setBackgroundColor(0)
    }
    alertDialogShower?.show(OpenCredits { view })
  }

  override fun onPreferenceTreeClick(preference: Preference): Boolean {
    if (preference.key.equals(PREF_CLEAR_ALL_HISTORY, ignoreCase = true)) {
      clearAllHistoryDialog()
    }
    if (preference.key.equals(PREF_CLEAR_ALL_NOTES, ignoreCase = true)) {
      showClearAllNotesDialog()
    }
    if (preference.key.equals(PREF_CREDITS, ignoreCase = true)) {
      openCredits()
    }
    if (preference.key.equals(PREF_EXPORT_BOOKMARK, ignoreCase = true) &&
      requestExternalStorageWritePermissionForExportBookmark()
    ) {
      showExportBookmarkDialog()
    }
    if (preference.key.equals(PREF_IMPORT_BOOKMARK, ignoreCase = true)) {
      showImportBookmarkDialog()
    }
    return true
  }

  @Suppress("NestedBlockDepth")
  private fun requestExternalStorageWritePermissionForExportBookmark(): Boolean {
    var isPermissionGranted = false
    if (sharedPreferenceUtil?.isPlayStoreBuildWithAndroid11OrAbove() == false &&
      Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU
    ) {
      if (requireActivity().checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        == PackageManager.PERMISSION_GRANTED
      ) {
        isPermissionGranted = true
      } else {
        storagePermissionForNotesLauncher?.launch(Manifest.permission.WRITE_EXTERNAL_STORAGE)
      }
    } else {
      isPermissionGranted = true
    }
    return isPermissionGranted
  }

  private var storagePermissionForNotesLauncher: ActivityResultLauncher<String>? =
    registerForActivityResult(
      ActivityResultContracts.RequestPermission()
    ) { isGranted ->
      if (isGranted) {
        // Successfully granted permission, so opening the export bookmark Dialog
        showExportBookmarkDialog()
      } else {
        if (shouldShowRequestPermissionRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
          /* shouldShowRequestPermissionRationale() returns false when:
             *  1) User has previously checked on "Don't ask me again", and/or
             *  2) Permission has been disabled on device
             */
          requireActivity().toast(
            R.string.ext_storage_permission_rationale_export_bookmark,
            Toast.LENGTH_LONG
          )
        } else {
          requireActivity().toast(
            R.string.ext_storage_write_permission_denied_export_bookmark,
            Toast.LENGTH_LONG
          )
          alertDialogShower?.show(
            KiwixDialog.ReadPermissionRequired,
            requireActivity()::navigateToAppSettings
          )
        }
      }
    }

  private fun showExportBookmarkDialog() {
    alertDialogShower?.show(
      KiwixDialog.YesNoDialog.ExportBookmarks,
      { libkiwixBookmarks?.exportBookmark() }
    )
  }

  private fun showImportBookmarkDialog() {
    alertDialogShower?.show(
      KiwixDialog.ImportBookmarks,
      ::showFileChooser
    )
  }

  private fun showFileChooser() {
    val intent = Intent().apply {
      action = Intent.ACTION_GET_CONTENT
      type = "*/*"
      addCategory(Intent.CATEGORY_OPENABLE)
    }
    try {
      fileSelectLauncher.launch(Intent.createChooser(intent, "Select a bookmark file"))
    } catch (ex: ActivityNotFoundException) {
      activity.toast(
        resources.getString(R.string.no_app_found_to_select_bookmark_file),
        Toast.LENGTH_SHORT
      )
    }
  }

  private val fileSelectLauncher =
    registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
      if (result.resultCode == RESULT_OK) {
        result.data?.data?.let { uri ->
          val contentResolver = requireActivity().contentResolver
          if (!isValidBookmarkFile(contentResolver.getType(uri))) {
            activity.toast(
              resources.getString(R.string.error_invalid_bookmark_file),
              Toast.LENGTH_SHORT
            )
            return@registerForActivityResult
          }

          createTempFile(contentResolver.openInputStream(uri)).apply {
            if (isValidXmlFile(this)) {
              CoroutineScope(Dispatchers.IO).launch {
                libkiwixBookmarks?.importBookmarks(this@apply)
              }
            } else {
              activity.toast(
                resources.getString(R.string.error_invalid_bookmark_file),
                Toast.LENGTH_SHORT
              )
            }
          }
        }
      }
    }

  private fun isValidXmlFile(file: File): Boolean {
    return try {
      DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(file)
      true
    } catch (ignore: Exception) {
      Log.e("IMPORT_BOOKMARKS", "Invalid XML file", ignore)
      false
    }
  }

  private fun createTempFile(inputStream: InputStream?): File {
    // create a temp file for importing the saved bookmarks
    val tempFile = File(requireActivity().externalCacheDir, "bookmark.xml")
    if (tempFile.exists()) {
      tempFile.delete()
    }
    tempFile.createNewFile()
    inputStream?.let {
      tempFile.outputStream().use(inputStream::copyTo)
    }
    return tempFile
  }

  private fun isValidBookmarkFile(mimeType: String?) =
    mimeType == "application/xml" || mimeType == "text/xml"

  @Suppress("NestedBlockDepth")
  fun onStorageDeviceSelected(storageDevice: StorageDevice) {
    lifecycleScope.launch {
      sharedPreferenceUtil?.let { sharedPreferenceUtil ->
        sharedPreferenceUtil.putPrefStorage(
          sharedPreferenceUtil.getPublicDirectoryPath(storageDevice.name)
        )
        sharedPreferenceUtil.putStoragePosition(
          if (storageDevice.isInternal) INTERNAL_SELECT_POSITION
          else EXTERNAL_SELECT_POSITION
        )
        setShowStorageOption()
        setStorage()
      }
    }
  }

  private fun setShowStorageOption() {
    sharedPreferenceUtil?.showStorageOption = false
  }

  companion object {
    const val PREF_VERSION = "pref_version"
    const val PREF_CLEAR_ALL_HISTORY = "pref_clear_all_history"
    const val PREF_CLEAR_ALL_NOTES = "pref_clear_all_notes"
    const val PREF_CREDITS = "pref_credits"
    const val PREF_PERMISSION = "pref_permissions"
    private const val ZOOM_OFFSET = 2
    private const val ZOOM_SCALE = 25
    private const val INTERNAL_TEXT_ZOOM = "text_zoom"
    private const val PREF_EXPORT_BOOKMARK = "pref_export_bookmark"
    private const val PREF_IMPORT_BOOKMARK = "pref_import_bookmark"
  }
}
